### ISSUE TITLE
Proxy to multiple providers

### ISSUE THREAD

**Issue by ywLBhGk2noZLbY7** [ncw/rclone#561]
Hi, sometimes when using rclone I want to upload it multiple different providers (OneDrive, Google Drive, etc...) for redundency, however, my upload speeds are terrible, so I have one of a few ways of doing this:-

1. Bite the bullet, upload to each provider on my terrible speed - issue being that it would take a _long_ time
2. Upload to one provider, then use a leased server to copy it between providers (`rclone provider1:/ provider2:/`) - issue being that it's not real time, I have to upload all my files to one provider before I can start cloning to other providers, also if my data is corrupt while uploading to the first provider I'm stuffed
3. Copying my files to a server, then using that to upload to different providers (rsync -> `rclone ${PATH} provider1:/` `rclone ${PATH} provider2:/`) - Issue being that I require HDD on the server (Something most cheap servers don't provide)

However, I was wondering if it'd be possible to implement a proxy/relay sort of system, where I host rclone on a remote server, and I upload my data to that rclone instance, however that rclone instance then "forks" my data and uploads it to X many providers. The benefit of this is:-

1. You get to do what you do in option 3, without the requirement for large amounts of HDD space on the server
2. Assuming your server's internet was fast enough, you could easily upload to multiple providers in real time (From the point-of-view of your home internet connection). For instance, imagine my upload speed was 5Mbit/s and my server had 100Mbit/s down and 100Mbit/s up. I could upload to my server, and my server could upload to 20 (100Mbit/s / 5Mbit/s) different providers in real-time without requiring my home connection to throttle itself at all.

I thought it'd be a good idea, I'm interested in other people's opinions.

**Comment by klauspost** [comment]
It is interesting, but probably should be in another application (rproxy, anyone?). It could of course piggyback on the cloud integrations of rclone.

The complexity of sync an app is not trivial. There are multiple things you need to take into consideration.

* Should it be completely async, or let the slowest backend set the speed?
* If it should buffer some operations - what happens if one backend returns an error?
* What if backends are desynced?
* Should multiple clients be able to write to the proxy at once?
* How should reads be handled? Desync issues galore.

So there are a lot of complex scenarios that have to be addressed. However, it is not an impossible task by any means.

I would personally go for option 2) in your case, and have some way of triggering upload on the remote server when your local upload has completed successfully. That way if your upload is incomplete, or your backup is broken you still have your previous backup on other providers.

**Comment by ncw** [comment]
I agree with @klauspost this is interesting but the whole proxy thing would be difficult to implement.  I think your option 2) is probably the most sensible.  Note that rclone checks the consistency of uploads, and you can use the `rclone check` command too if you are worried about data consistency.

I've been playing with [minio](https://minio.io/) recently (an easy to install S3 server) which works with rclone (with the `--size-only` flag) - that could be part of a solution too.

**Comment by mholt** [comment]
@ncw I can understand how proxying to multiple backends is difficult, but assuming there's just a single destination: can rclone copy to a remote rclone which copies to the cloud in a streaming fashion? To avoid having to store a copy of all the data on the proxy server...  (c.f. https://forum.rclone.org/t/proxying-through-a-remote-rclone-instance/5318?u=matt)
